internal/v1/handler.go

```diff
@@ -498,6 +498,89 @@ func convertIgnoreImageTypeToSlice(ignoreImageTypes *[]ImageTypes) []string {
 	return stringSlice
 }
 
+func newComposeResponse(data []ComposesResponseItem, count int, first, last string) ComposesResponse {
+	return ComposesResponse{
+		Data: data,
+		Meta: struct {
+			Count int `json:"count"`
+		}{
+			count,
+		},
+		Links: struct {
+			First string `json:"first"`
+			Last  string `json:"last"`
+		}{first, last},
+	}
+}
+
+func (h *Handlers) getComposesResponse(specVersion string, orgId string, since time.Duration, limit, offset int, ignoreImageTypeStrings []string) (ComposesResponse, error) {
+	composes, count, err := h.server.db.GetComposes(orgId, since, limit, offset, ignoreImageTypeStrings)
+	if err != nil {
+		return ComposesResponse{}, err
+	}
+
+	data := []ComposesResponseItem{}
+	for _, c := range composes {
+		var cmpr ComposeRequest
+		err = json.Unmarshal(c.Request, &cmpr)
+		if err != nil {
+			return ComposesResponse{}, err
+		}
+		data = append(data, ComposesResponseItem{
+			CreatedAt: c.CreatedAt.Format(time.RFC3339),
+			Id:        c.Id,
+			ImageName: c.ImageName,
+			Request:   cmpr,
+			ClientId:  (*ClientId)(c.ClientId),
+		})
+	}
+
+	lastOffset := count - 1
+	if lastOffset < 0 {
+		lastOffset = 0
+	}
+	first := fmt.Sprintf("%v/v%v/composes?offset=0&limit=%v", RoutePrefix(), specVersion, limit)
+	last := fmt.Sprintf("%v/v%v/composes?offset=%v&limit=%v", RoutePrefix(), specVersion, lastOffset, limit)
+	return newComposeResponse(data, count, first, last), nil
+}
+
+func (h *Handlers) getBlueprintComposesResponse(specVersion string, orgId string, blueprintId uuid.UUID, blueprintVersion *int, since time.Duration, limit, offset int, ignoreImageTypeStrings []string) (ComposesResponse, error) {
+	composes, count, err := h.server.db.GetBlueprintComposes(orgId, blueprintId, since, limit, offset, ignoreImageTypeStrings)
+	if err != nil {
+		return ComposesResponse{}, err
+	}
+
+	data := make([]ComposesResponseItem, 0, len(composes))
+	for _, c := range composes {
+		var cmpr ComposeRequest
+		err = json.Unmarshal(c.Request, &cmpr)
+		if err != nil {
+			return ComposesResponse{}, err
+		}
+		data = append(data, ComposesResponseItem{
+			BlueprintId:      &c.BlueprintId,
+			BlueprintVersion: &c.BlueprintVersion,
+			CreatedAt:        c.CreatedAt.Format(time.RFC3339),
+			Id:               c.Id,
+			ImageName:        c.ImageName,
+			Request:          cmpr,
+			ClientId:         (*ClientId)(c.ClientId),
+		})
+	}
+
+	lastOffset := count - 1
+	if lastOffset < 0 {
+		lastOffset = 0
+	}
+	path := fmt.Sprintf("%v/v%v/composes?blueprint_id=%v", RoutePrefix(), specVersion, blueprintId.String())
+	if blueprintVersion != nil {
+		path += fmt.Sprintf("&blueprint_version=%d", *blueprintVersion)
+	}
+	first := path + fmt.Sprintf("&offset=0&limit=%v", limit)
+	last := path + fmt.Sprintf("&offset=%v&limit=%v", lastOffset, limit)
```

Comment:
I really think that path construction should be part of the helper function(s), this is too much to ask for a user who will be copy-pasting a new handler 8 months from now and not messing this up. Just pass a path portion into a function and it should do all the magic.